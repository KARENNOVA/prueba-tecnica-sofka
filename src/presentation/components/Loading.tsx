import React from 'react'
import { ActivityIndicator, StyleSheet, View } from 'react-native'
import { Text } from 'react-native-gesture-handler'

export const Loading = () => {
    return (
        <View style={styles.loaderContainer}>
            <ActivityIndicator size="large" color="#6200ea" />
            <Text>Cargando...</Text>
        </View>
    )
}

const styles = StyleSheet.create({
    loaderContainer: {
      flex: 1,
      justifyContent: 'center',
      alignItems: 'center',
    },
    container: {
      flex: 1,
      justifyContent: 'center',
      alignItems: 'center',
    },
  });
