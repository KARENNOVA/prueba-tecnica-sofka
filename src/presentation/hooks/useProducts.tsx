import { useEffect, useState } from 'react';
import type { Product } from '../../core/entities/product.entity';

import * as UseCases from '../../core/use-cases';
import { productFetcher } from '../../config/adapters/product.adapter';

interface Props {
  type: 'create' | 'edit' | 'delete' | 'list' | 'get'
  id?: string;
}

export const useProducts = ({ type, id }: Props) => {

  const [isLoading, setIsLoading] = useState(true);
  const [listProducts, setListProducts] = useState<Product[]>([]);
  const [product, setProduct] = useState<Product>({
    id: "",
    nombre: "",
    descripcion: "",
    logo: "",
    fechaLiberacion: "",
    fechaRevision: "",
  });
  const [changeStatus, setChangeStatus] = useState(false);

  useEffect(() => {
    if (type === 'list')
      fetchProducts()
  }, [changeStatus])

  useEffect(() => {
    if (type === 'get' && id) {
      fetchProductById(id)
    }
  }, [id])


  const fetchProducts = async (): Promise<Product[]> => {
    setIsLoading(true)
    const listProd = await UseCases.listProductsUseCase(productFetcher);
    setListProducts(listProd);
    setIsLoading(false);
    return listProd
  };


  const fetchProductById = async (productId: string): Promise<Product> => {
    setIsLoading(true)
    const product = await UseCases.getProductByIdUseCase(productFetcher, productId);
    setProduct(product)
    setIsLoading(false)
    return product
  };

  const createProduct = async (product: Product): Promise<void> => {
    const createdProduct = await UseCases.createProductUseCase(productFetcher, product);
    // setListProducts(prev => [...prev, createdProduct]); // actualiza el estado
    setChangeStatus(!changeStatus)
  };

  const updateProduct = async (productId: string, product: Product): Promise<void> => {
    const updatedProduct = await UseCases.updateProductUseCase(productFetcher, productId, product);
    // setListProducts(prev => prev.map(p => p.id === productId ? updatedProduct : p));
    setChangeStatus(!changeStatus)
  };

  const deleteProduct = async (productId: string): Promise<void> => {
    await UseCases.deleteProductUseCase(productFetcher, productId);
    // setListProducts(prev => prev.filter(p => p.id !== productId));
    setChangeStatus(!changeStatus)
  };

  // ðŸš€ Eliminar producto
  const verifyProduct = async (productId: string): Promise<boolean> => {
    return await UseCases.getVerifyProductUseCase(productFetcher, productId);
  };

  return {
    isLoading,
    listProducts,
    product,
    fetchProducts,
    fetchProductById,
    createProduct,
    updateProduct,
    deleteProduct,
    verifyProduct
  };
};
