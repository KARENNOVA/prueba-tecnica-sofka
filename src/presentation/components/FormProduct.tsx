import React from 'react'
import { View, Text, TextInput, StyleSheet, Pressable } from 'react-native';
import { useForm, Controller } from 'react-hook-form';
import { globalStyles } from '../theme/global.styles';
import { Product } from '../../core/entities/product.entity';
import { useProducts } from '../hooks/useProducts';
import { NavigationProp, useNavigation } from '@react-navigation/native';
import { RootStackParams } from '../routes/StackNavigation';

interface Props {
    type: 'create' | 'edit',
    product?: Product
}

export const FormProduct = ({ type, product }: Props) => {
    const navigation = useNavigation<NavigationProp<RootStackParams>>();
    const { control, handleSubmit, setValue, formState: { errors }, reset } = useForm({
        defaultValues: {
            id: '',
            nombre: '',
            descripcion: '',
            logo: '',
            fechaLiberacion: '',
            fechaRevision: '',
        },
        values: product
    });


    const {
        createProduct,
        updateProduct
    } = useProducts({ type: 'create' });

    const esFechaValida = (fecha: any) => {
        const regex = /^\d{2}\/\d{2}\/\d{4}$/; // formato dd/mm/yyyy
        if (!regex.test(fecha)) return 'Formato inválido (dd/mm/yyyy)';

        const [day, month, year] = fecha.split('/').map(Number);
        const fechaObj = new Date(year, month - 1, day);

        const hoy = new Date();
        hoy.setHours(0, 0, 0, 0);

        if (isNaN(fechaObj.getTime())) return 'Fecha inválida';
        if (fechaObj < hoy) return 'La fecha debe ser igual o mayor a hoy';

        return true;
    };

    const calcularFechaRevision = (fecha: any) => {
        const [day, month, year] = fecha.split('/').map(Number);
        const fechaRevision = new Date(year + 1, month - 1, day);
        return fechaRevision.toLocaleDateString('es-ES');
    };

    const handleFechaLiberacionChange = (fecha: any) => {
        setValue('fechaLiberacion', fecha, { shouldValidate: true });

        if (esFechaValida(fecha) === true) {
            const revision = calcularFechaRevision(fecha);
            setValue('fechaRevision', revision);
        } else {
            setValue('fechaRevision', '');
        }
    };

    const onSubmit = (data: any) => {
        try {
            if (type === 'create') {
                createProduct(data);
            } else {
                updateProduct(data.id, data)
            }
            reset();
            navigation.navigate('Products');
        } catch (error) {
            console.error(error)
        }
    };

    return (
        <>
            <Text style={styles.label}>ID</Text>
            <Controller
                control={control}
                name="id"
                rules={{ 
                    required: 'El ID es obligatorio', 
                    minLength: { value: 3, message: 'Debe tener al menos 3 caracteres' },
                    maxLength: { value: 10, message: 'Debe tener máximo 10 caracteres' }
                }}
                
                render={({ field: { onChange, value } }) => (
                    <TextInput
                        style={[styles.input, errors.id && styles.inputError]}
                        value={value}
                        onChangeText={onChange}
                        placeholder="Ingresa el ID"
                    />
                )}
            />
            {errors.id && <Text style={styles.errorText}>{errors.id.message}</Text>}

            <Text style={styles.label}>Nombre</Text>
            <Controller
                control={control}
                name="nombre"
                rules={{ 
                    required: 'El nombre es obligatorio', 
                    minLength: { value: 5, message: 'Debe tener al menos 3 caracteres' },
                    maxLength: { value: 100, message: 'Debe tener máximo 50 caracteres' }
                }}
                render={({ field: { onChange, value } }) => (
                    <TextInput
                        style={[styles.input, errors.nombre && styles.inputError]}
                        value={value}
                        onChangeText={onChange}
                        placeholder="Ingresa el nombre"
                    />
                )}
            />
            {errors.nombre && <Text style={styles.errorText}>{errors.nombre.message}</Text>}

            <Text style={styles.label}>Descripción</Text>
            <Controller
                control={control}
                name="descripcion"
                rules={{ 
                    required: 'La descripción es obligatoria', 
                    minLength: { value: 10, message: 'Debe tener al menos 3 caracteres' },
                    maxLength: { value: 200, message: 'Debe tener máximo 50 caracteres' }
                }}
                render={({ field: { onChange, value } }) => (
                    <TextInput
                        style={[styles.input, errors.descripcion && styles.inputError]}
                        value={value}
                        onChangeText={onChange}
                        placeholder="Ingresa la descripción"
                    />
                )}
            />
            {errors.descripcion && <Text style={styles.errorText}>{errors.descripcion.message}</Text>}

            <Text style={styles.label}>Logo</Text>
            <Controller
                control={control}
                name="logo"
                rules={{
                    required: 'El logo es obligatorio',
                    // pattern: {
                    //     value: /^https?:\/\/.*\.(?:png|jpg|jpeg|svg|gif|webp)$/,
                    //     message: 'Debe ser una URL válida de imagen'
                    // }
                }}
                render={({ field: { onChange, value } }) => (
                    <TextInput
                        style={[styles.input, errors.logo && styles.inputError]}
                        value={value}
                        onChangeText={onChange}
                        placeholder="URL del logo"
                    />
                )}
            />
            {errors.logo && <Text style={styles.errorText}>{errors.logo.message}</Text>}

            <Text style={styles.label}>Fecha Liberación</Text>
            <Controller
                control={control}
                name="fechaLiberacion"
                rules={{
                    required: 'Campo requerido',
                    validate: esFechaValida
                }}
                render={({ field: { value } }) => (
                    <TextInput
                        style={styles.input}
                        value={value}
                        onChangeText={handleFechaLiberacionChange}
                        placeholder="dd/mm/yyyy"
                        keyboardType="numeric"
                    />
                )}
            />
            {errors.fechaLiberacion && <Text style={styles.error}>{errors.fechaLiberacion.message}</Text>}

            <Text style={styles.label}>Fecha Revisión</Text>
            <Controller
                control={control}
                name="fechaRevision"
                rules={{ required: 'Campo requerido' }}
                render={({ field: { value } }) => (
                    <TextInput
                        style={[styles.input, styles.disabledInput]}
                        value={value}
                        editable={false}
                    />
                )}
            />
            {errors.fechaRevision && <Text style={styles.error}>{errors.fechaRevision.message}</Text>}

            <View style={styles.buttonContainer}>
                <Pressable
                    onPress={handleSubmit(onSubmit)}
                    style={globalStyles.prymaryButton}
                >
                    <Text style={globalStyles.prymaryButtonText}>{type=== "create" ? 'Agregar' : 'Editar'}</Text>
                </Pressable>
                <Pressable
                    onPress={() => reset()}
                    style={globalStyles.secondaryButton}
                >
                    <Text style={globalStyles.secondaryButtonText}>Reiniciar</Text>
                </Pressable>
            </View>
        </>
    );
}

const styles = StyleSheet.create({
    input: {
        borderWidth: 1,
        borderColor: '#ccc',
        padding: 10,
        marginBottom: 10,
        borderRadius: 5,
    },
    inputError: {
        borderColor: 'red',
    },
    errorText: {
        color: 'red',
        marginBottom: 10,
    },
    buttonContainer: {
        marginTop: 20,
    },
    label: {
        fontWeight: 'bold',
        marginBottom: 5,
    },
    disabledInput: {
        backgroundColor: '#f0f0f0',
        color: '#888',
    },
    error: {
        color: 'red',
        fontSize: 12,
    },
});





