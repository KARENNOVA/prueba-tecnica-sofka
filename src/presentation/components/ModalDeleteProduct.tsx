import React, { useState } from 'react'
import { Modal, Pressable, StyleSheet, Text, View } from 'react-native';
import { globalStyles } from '../theme/global.styles';
import { Product } from '../../core/entities/product.entity';
import { useProducts } from '../hooks/useProducts';
import { NavigationProp, useNavigation } from '@react-navigation/native';
import { RootStackParams } from '../routes/StackNavigation';

interface Props {
    product: Product
}

export const ModalDeleteProduct = ({ product }: Props) => {
    const navigation = useNavigation<NavigationProp<RootStackParams>>();
    const [modalVisible, setModalVisible] = useState(false);
    const {
        deleteProduct
    } = useProducts({ type: 'delete' });

    const handleDeletePress = () => {
        setModalVisible(true);
    };

    const handleConfirmDelete = () => {
        // Aquí ejecutas la lógica para eliminar
        deleteProduct(product.id)
        setModalVisible(false);
        navigation.navigate('Products');
    };
    return (
        <>
            <Pressable onPress={handleDeletePress} style={styles.deleteButton}>
                <Text style={styles.deleteButtonText}>Eliminar</Text>
            </Pressable>
            <Modal
                transparent
                visible={modalVisible}
                animationType="slide"
                onRequestClose={() => setModalVisible(false)}
            >
                <View style={styles.modalOverlay}>
                    <View style={styles.modalContainer}>
                        {/* Cerrar modal */}
                        <Pressable
                            style={styles.closeButton}
                            onPress={() => setModalVisible(false)}
                        >
                            <Text style={styles.closeButtonText}>✕</Text>
                        </Pressable>

                        {/* Contenido */}
                        <Text style={styles.modalText}>
                            ¿Estás seguro de eliminar el producto {product.nombre}?
                        </Text>

                        {/* Botones */}
                        <Pressable
                            style={globalStyles.prymaryButton}
                            onPress={handleConfirmDelete}
                        >
                            <Text style={globalStyles.prymaryButtonText}>Confirmar</Text>
                        </Pressable>

                        <Pressable
                            style={globalStyles.secondaryButton}
                            onPress={() => setModalVisible(false)}
                        >
                            <Text style={globalStyles.secondaryButtonText}>Cancelar</Text>
                        </Pressable>
                    </View>
                </View>
            </Modal>
        </>
    )
}

const styles = StyleSheet.create({
    container: {
        flex: 1,
        padding: 20,
        backgroundColor: '#fff',
        alignItems: 'center',
    },
    bankTitle: {
        fontSize: 18,
        fontWeight: 'bold',
        marginBottom: 10,
    },
    idText: {
        fontSize: 16,
        fontWeight: '600',
        marginBottom: 5,
    },
    extraInfo: {
        fontSize: 14,
        color: 'gray',
        marginBottom: 20,
    },
    deleteButton: {
        width: '100%',
        paddingVertical: 12,
        backgroundColor: '#E53935',
        alignItems: 'center',
        borderRadius: 8,
        marginTop: 20,
    },
    deleteButtonText: {
        fontSize: 16,
        color: '#fff',
        fontWeight: 'bold',
    },
    modalOverlay: {
        flex: 1,
        backgroundColor: 'rgba(0,0,0,0.5)',
        justifyContent: 'flex-end',
    },
    modalContainer: {
        backgroundColor: '#fff',
        padding: 20,
        borderTopLeftRadius: 16,
        borderTopRightRadius: 16,
    },
    closeButton: {
        alignSelf: 'flex-end',
        marginBottom: 10,
    },
    closeButtonText: {
        fontSize: 18,
        fontWeight: 'bold',
        color: '#333',
    },
    modalText: {
        fontSize: 16,
        textAlign: 'center',
        marginBottom: 20,
    },
    actionButton: {
        paddingVertical: 12,
        borderRadius: 8,
        alignItems: 'center',
        marginBottom: 10,
    },
    confirmButton: {
        backgroundColor: '#FFD700',
    },
    confirmText: {
        fontWeight: 'bold',
        color: '#000',
    },
    cancelButton: {
        backgroundColor: '#E0E0E0',
    },
    cancelText: {
        fontWeight: 'bold',
        color: '#000',
    },
});