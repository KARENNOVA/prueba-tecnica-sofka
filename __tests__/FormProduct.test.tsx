import React from 'react';
import renderer, { act } from 'react-test-renderer';

import { NavigationContainer } from '@react-navigation/native';
import { TextInput, Button } from 'react-native';
import { useProducts } from '../src/presentation/hooks/useProducts';
import { FormProduct } from '../src/presentation/components/FormProduct';

// Mock de `useProducts`
jest.mock('../src/presentation/hooks/useProducts', () => ({
    useProducts: jest.fn()
}));

const mockCreateProduct = jest.fn();
const mockUpdateProduct = jest.fn();
const mockVerifyProduct = jest.fn().mockResolvedValue(false);

beforeEach(() => {
    (useProducts as jest.Mock).mockReturnValue({
        createProduct: mockCreateProduct,
        updateProduct: mockUpdateProduct,
        verifyProduct: mockVerifyProduct,
    });
});

describe('FormProduct Screen', () => {

    const renderWithNavigation = (props: any) => {
        return renderer.create(
            <NavigationContainer>
                <FormProduct {...props} />
            </NavigationContainer>
        );
    };

    test('renderiza correctamente en modo creación', () => {
        let tree;
        act(() => {
            tree = renderWithNavigation({ type: 'create' }).toJSON();
        });
        expect(tree).toMatchSnapshot(); // Snapshot para validar estructura inicial
    });

    test('muestra error si se envía sin ID', () => {
        const testRenderer = renderWithNavigation({ type: 'create' });

        act(() => {
            const button = testRenderer.root.findByProps({ title: 'Agregar' });
            button.props.onPress();
        });

        const errorMessage = testRenderer.root.findAllByProps({ children: 'El ID es obligatorio' });
        expect(errorMessage.length).toBe(1); // Debería aparecer el mensaje de error
    });

    test('valida formato de fecha incorrecto', () => {
        const testRenderer = renderWithNavigation({ type: 'create' });

        const fechaInput = testRenderer.root.findByProps({ placeholder: 'dd/mm/yyyy' });
        act(() => fechaInput.props.onChangeText('2025-02-30'));

        const button = testRenderer.root.findByProps({ title: 'Agregar' });
        act(() => button.props.onPress());

        const errorMessage = testRenderer.root.findAllByProps({ children: 'Formato inválido (dd/mm/yyyy)' });
        expect(errorMessage.length).toBe(1);
    });

    test('crea un producto correctamente', async () => {
        const testRenderer = renderWithNavigation({ type: 'create' });

        act(() => {
            testRenderer.root.findByProps({ placeholder: 'Ingresa el ID' }).props.onChangeText('123');
            testRenderer.root.findByProps({ placeholder: 'Ingresa el nombre' }).props.onChangeText('Producto Test');
            testRenderer.root.findByProps({ placeholder: 'Ingresa la descripción' }).props.onChangeText('Descripción válida');
            testRenderer.root.findByProps({ placeholder: 'URL del logo' }).props.onChangeText('https://example.com/logo.png');
            testRenderer.root.findByProps({ placeholder: 'dd/mm/yyyy' }).props.onChangeText('28/02/2025');
        });

        const button = testRenderer.root.findByProps({ title: 'Agregar' });
        await act(async () => {
            button.props.onPress();
        });

        expect(mockCreateProduct).toHaveBeenCalledWith({
            id: '123',
            nombre: 'Producto Test',
            descripcion: 'Descripción válida',
            logo: 'https://example.com/logo.png',
            fechaLiberacion: '28/02/2025',
            fechaRevision: '28/02/2026' // asumimos que el año se suma automáticamente
        });
    });

    test('no permite crear producto si el ID ya existe', async () => {
        mockVerifyProduct.mockResolvedValueOnce(true);

        const testRenderer = renderWithNavigation({ type: 'create' });

        const idInput = testRenderer.root.findByProps({ placeholder: 'Ingresa el ID' });

        await act(async () => {
            idInput.props.onChangeText('123');
            idInput.props.onBlur(); // Disparar la validación
        });

        const errorMessage = testRenderer.root.findAllByProps({ children: 'Este ID ya existe' });
        expect(errorMessage.length).toBe(1);
    });

});
