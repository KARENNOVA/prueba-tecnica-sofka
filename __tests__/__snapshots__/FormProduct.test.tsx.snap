// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormProduct Screen renderiza correctamente en modo creaci√≥n 1`] = `null`;
